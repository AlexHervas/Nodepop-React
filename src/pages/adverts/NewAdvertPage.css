.newAdvertPage-form {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  max-width: 450px;
  margin: auto;
  padding: 2rem;
  background-color: #4a4a4a;
  border-radius: 10px;
  box-shadow: 0 4px 10px rgba(0, 0, 0, 0.2);
}

.newAdvertPage-form label {
  color: #9a9a9a;
  font-weight: bold;
  font-size: 0.9rem;
}

.error {
  color: #ff4a4a;
  background: rgba(255, 74, 74, 0.1);
  padding: 0.5rem;
  border-radius: 5px;
  text-align: center;
}

.button {
  background-color: #7a7a7a;
  color: white;
  padding: 0.75rem;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  font-size: 1rem;
  transition:
    background 0.3s,
    transform 0.2s;
}

.button:hover {
  background-color: #9a9a9a;
  transform: scale(1.02);
}

.button:disabled {
  background-color: #3a3a3a;
  cursor: not-allowed;
  opacity: 0.6;
}

@media (max-width: 500px) {
  .newAdvertPage-form {
    max-width: 100%;
    padding: 1.5rem;
  }
}
