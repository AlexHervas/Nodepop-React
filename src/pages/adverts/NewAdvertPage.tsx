import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import Page from "../../components/layout/Page";
import FormField from "../../components/shared/FormField";
import Button from "../../components/shared/Button";
import "./NewAdvertPage.css";

// Importamos los hooks tipados y las acciones/thunks del store
import { useAppDispatch, useAppSelector } from "../../store";
import { advertsCreate, tagsLoaded } from "../../store/actions";
import { isApiClientError } from "../../api/client";

// Importamos el tipo para el nuevo anuncio
import { NewAdvert } from "./types";

function NewAdvertPage() {
  // Estado local para el nuevo anuncio
  const [advert, setAdvert] = useState<NewAdvert>({
    name: "",
    sale: true,
    price: 0,
    tags: [],
    photo: null,
  });
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(false);

  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  // Obtenemos las opciones de tags desde el store
  const tagOptions = useAppSelector((state) => state.adverts.tags);
  // Si el array de tags está vacío, consideramos que aún se están cargando
  const loadingTags = tagOptions.length === 0;

  // Al montar, se despacha la acción para cargar los tags globalmente
  useEffect(() => {
    dispatch(tagsLoaded() as any);
  }, [dispatch]);

  // Maneja el envío del formulario para crear el anuncio
  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setErrorMessage(null);
    setIsLoading(true);

    // Preparamos los datos en FormData para enviar, ya que el servicio espera FormData
    const formData = new FormData();
    formData.append("name", advert.name);
    formData.append("sale", String(advert.sale));
    formData.append("price", String(advert.price));
    formData.append("tags", advert.tags.join(","));
    if (advert.photo) {
      formData.append("photo", advert.photo);
    }

    try {
      // Despachamos la acción asíncrona para crear el anuncio.
      // Se espera que advertsCreate retorne el anuncio creado.
      const createdAdvert = await dispatch(advertsCreate(formData) as any);
      navigate(`/adverts/${createdAdvert.id}`);
    } catch (error) {
      if (isApiClientError(error)) {
        if (error.code === "UNAUTHORIZED") {
          navigate("/login");
        } else {
          setErrorMessage(
            error.message || "An error occurred. Please try again.",
          );
        }
      }
    } finally {
      setIsLoading(false);
    }
  };

  // Maneja cambios en los inputs del formulario
  const handleChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>,
  ) => {
    const { name, value, type, files, checked } =
      event.target as HTMLInputElement;

    setAdvert((prevAdvert) => {
      if (type === "file" && files) {
        return { ...prevAdvert, photo: files[0] };
      }
      if (type === "checkbox") {
        // Actualiza el array de tags
        const newTags = checked
          ? [...prevAdvert.tags, value]
          : prevAdvert.tags.filter((tag) => tag !== value);
        return { ...prevAdvert, tags: newTags };
      }
      return {
        ...prevAdvert,
        [name]: name === "sale" ? value === "true" : value,
      };
    });
  };

  const { name, price, tags } = advert;
  // Deshabilita el botón si faltan campos obligatorios o si está en proceso
  const isDisabled = !name || price <= 0 || tags.length === 0 || isLoading;

  return (
    <Page title="Create your advert">
      <form className="newAdvertPage-form" onSubmit={handleSubmit}>
        {/* Campo para el nombre */}
        <FormField
          type="text"
          name="name"
          label="Name of your item"
          className="name"
          value={advert.name}
          onChange={handleChange}
          required
        />

        {/* Opciones para seleccionar si se vende o se compra */}
        {["true", "false"].map((value) => (
          <FormField
            key={value}
            type="radio"
            name="sale"
            label={value === "true" ? "Sell" : "Buy"}
            className="sale"
            value={value}
            onChange={handleChange}
            checked={advert.sale.toString() === value}
          />
        ))}

        {/* Campo para ingresar el precio */}
        <FormField
          type="number"
          name="price"
          label="Price of your item"
          className="price"
          value={advert.price}
          onChange={handleChange}
          required
        />

        {/* Sección para seleccionar etiquetas */}
        <div className="formField">
          <span>Tags</span>
          {loadingTags && <p>Loading tags...</p>}
          {!loadingTags &&
            tagOptions.map((tag) => (
              <label key={tag}>
                <input
                  type="checkbox"
                  name="tags"
                  value={tag}
                  checked={advert.tags.includes(tag)}
                  onChange={handleChange}
                />
                {tag.charAt(0).toUpperCase() + tag.slice(1)}
              </label>
            ))}
        </div>

        {/* Campo para subir foto */}
        <FormField
          type="file"
          name="photo"
          label="Photo"
          className="photo"
          onChange={handleChange}
        />

        {/* Mensaje de error */}
        {errorMessage && <div className="error">{errorMessage}</div>}

        {/* Botón para enviar */}
        <Button
          className="button"
          type="submit"
          $variant="primary"
          disabled={isDisabled}
        >
          {isLoading ? "Creating..." : "Create Advert"}
        </Button>
      </form>
    </Page>
  );
}

export default NewAdvertPage;
