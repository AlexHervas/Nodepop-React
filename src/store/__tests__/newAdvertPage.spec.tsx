import renderer from "react-test-renderer";
import { describe, it, expect } from "vitest";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import NewAdvertPage from "../../pages/adverts/NewAdvertPage";
import configureStore from "../../store";
import { createBrowserRouter } from "react-router-dom";

// Creamos un router y un store para el test.
// Aseg√∫rate de que la estructura del estado coincide con la que espera tu store.
const router = createBrowserRouter([{ path: "*", element: <NewAdvertPage /> }]);
const initialState = {
  auth: false,
  adverts: { data: [], loaded: true, tags: ["tag1", "tag2"] },
  ui: { pending: false, error: null },
};
const store = configureStore(initialState, router);

describe("NewAdvertPage Snapshot", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <NewAdvertPage />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
