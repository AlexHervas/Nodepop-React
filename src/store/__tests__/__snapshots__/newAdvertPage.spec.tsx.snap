// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewAdvertPage Snapshot > renders correctly 1`] = `
<div
  className="layout-page"
>
  <h2
    className="layout-title"
  >
    Create your advert
  </h2>
  <form
    className="newAdvertPage-form"
    onSubmit={[Function]}
  >
    <div
      className="formField"
    >
      <label
        className="formField-label"
      >
        <span>
          Name of your item
        </span>
        <input
          autoComplete="off"
          className="name"
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="formField"
    >
      <label
        className="formField-label"
      >
        <span>
          Sell
        </span>
        <input
          autoComplete="off"
          checked={true}
          className="sale"
          name="sale"
          onChange={[Function]}
          type="radio"
          value="true"
        />
      </label>
    </div>
    <div
      className="formField"
    >
      <label
        className="formField-label"
      >
        <span>
          Buy
        </span>
        <input
          autoComplete="off"
          checked={false}
          className="sale"
          name="sale"
          onChange={[Function]}
          type="radio"
          value="false"
        />
      </label>
    </div>
    <div
      className="formField"
    >
      <label
        className="formField-label"
      >
        <span>
          Price of your item
        </span>
        <input
          autoComplete="off"
          className="price"
          name="price"
          onChange={[Function]}
          required={true}
          type="number"
          value={0}
        />
      </label>
    </div>
    <div
      className="formField"
    >
      <span>
        Tags
      </span>
      <label>
        <input
          checked={false}
          name="tags"
          onChange={[Function]}
          type="checkbox"
          value="tag1"
        />
        Tag1
      </label>
      <label>
        <input
          checked={false}
          name="tags"
          onChange={[Function]}
          type="checkbox"
          value="tag2"
        />
        Tag2
      </label>
    </div>
    <div
      className="formField"
    >
      <label
        className="formField-label"
      >
        <span>
          Photo
        </span>
        <input
          autoComplete="off"
          className="photo"
          name="photo"
          onChange={[Function]}
          type="file"
        />
      </label>
    </div>
    <button
      className="sc-FEMpB fhnnRn button"
      disabled={true}
      type="submit"
    >
      Create Advert
    </button>
  </form>
</div>
`;
